src/compiler/checker.ts

```diff
@@ -4053,23 +4364,79 @@ namespace ts {
             return getDeclaredTypeOfSymbol(symbol);
         }
 
-        function getTypeFromTypeReference(node: TypeReferenceNode | ExpressionWithTypeArguments): Type {
+        function getTypeReferenceName(node: TypeReferenceNode | ExpressionWithTypeArguments | JSDocTypeReference): LeftHandSideExpression | EntityName {
+            switch (node.kind) {
+                case SyntaxKind.TypeReference:
+                    return (<TypeReferenceNode>node).typeName;
+                case SyntaxKind.JSDocTypeReference:
+                    return (<JSDocTypeReference>node).name;
+                case SyntaxKind.ExpressionWithTypeArguments:
+                    // We only support expressions that are simple qualified names. For other
+                    // expressions this produces undefined.
+                    if (isSupportedExpressionWithTypeArguments(<ExpressionWithTypeArguments>node)) {
+                        return (<ExpressionWithTypeArguments>node).expression;
+                    }
+
+                // fall through;
+            }
+
+            return undefined;
+        }
+
+        function resolveTypeReferenceName(
+            node: TypeReferenceNode | ExpressionWithTypeArguments | JSDocTypeReference,
+            typeReferenceName: LeftHandSideExpression | EntityName) {
+
+            if (!typeReferenceName) {
+                return unknownSymbol;
+            }
+
+            let symbol = resolveEntityName(typeReferenceName, SymbolFlags.Type);
+            if (!symbol && node.kind === SyntaxKind.JSDocTypeReference) {
+                // If the reference didn't resolve to a type, try seeing if results to a 
+                // value.  If it does, get the type of that value.
+                symbol = resolveEntityName(typeReferenceName, SymbolFlags.Value);

```

Comment:
The current logic was picked simply because it seemed to work "good enough" for the set of files i tested against.  In general, we wanted to match against the type.  But in a lot of cases where the type was not found, falling back to a value lookup was sufficient.  Feel free to change/adjust this logic if you feel like another heuristic will be better for real world code out there. :)